README.md

```diff
@@ -118,26 +118,250 @@ services:
+
+### Changes and tasks
+
+When Pebble performs a (potentially invasive or long-running) operation such as starting or stopping a service, it records a "change" object with one or more "tasks" in it. The daemon records this state in a JSON file on disk at `$PEBBLE/.pebble.state`.
+
+To see recent changes, for this or previous server runs, use `pebble changes`. You might see something like this:
+
+```
+$ pebble changes
+ID  Status  Spawn                Ready                Summary
+1   Done    today at 14:33 NZDT  today at 14:33 NZDT  Autostart service "srv1"
+2   Done    today at 15:26 NZDT  today at 15:26 NZDT  Start service "srv2"
+3   Done    today at 15:26 NZDT  today at 15:26 NZDT  Stop service "srv1" and 1 more
+```
+
+To drill down and see the tasks that make up a change, use `pebble tasks <change-id>`:
+
+```
+$ pebble tasks 3
+Status  Spawn                Ready                Summary
+Done    today at 15:26 NZDT  today at 15:26 NZDT  Stop service "srv1"
+Done    today at 15:26 NZDT  today at 15:26 NZDT  Stop service "srv2"
+```
+
+### Exec (one-shot commands)
+
+Pebble's "exec" feature allows you to run arbitrary commands on the server. This is intended for short-running programs; the processes started with exec don't use the service manager.
+
+For example, you could use `exec` to run pg_dump and create a PostgreSQL database backup:
+
+```
+$ pebble exec pg_dump mydb
+--
+-- PostgreSQL database dump
+--
+...
+```
+
+The exec feature uses WebSockets under the hood, and allows you to stream stdin to the process, as well as stream stdout and stderr back. When running `pebble exec`, you can specify the working directory to run in (`-w`), environment variables to set (`--env`), and the user and group to run as (`--uid`/`--user` and `--gid`/`--group`).
+
+You can also apply a timeout with `--timeout`, for example:
+
+```
+$ pebble exec --timeout 1s -- sleep 3
+error: cannot perform the following tasks:
+- exec command "sleep" (timed out after 1s: context deadline exceeded)
+```
+
+### File management
+
+Pebble provides various API calls and commands to manage files and directories on the server. The simplest way to use these is with the commands below, several of which should be familiar:
+
+```
+$ pebble ls <path>              # list file information (like "ls")
+$ pebble mkdir <path>           # create a directory (like "mkdir")
+
+# TODO -- the following commands are coming soon
+$ pebble rm <path>              # remove a file or directory (like "rm")
+$ pebble push <local> <remote>  # copy file to server (like "cp")
+$ pebble pull <remote> <local>  # copy file from server (like "cp")
+```
 
 ## Layer specification
 
+Below is the full specification for a Pebble configuration layer. Layers are added statically using a file in `$PEBBLE/layers`, or dynamically via the layers API or `pebble add`.
```

Comment:
Seems a bit weird to cover so many aspects of Pebble without even mentioning how to set it up at all. It's impossible to use Pebble before reading the section below.

We might have a trivial config close to the top, to enable experimentation with the remaining commands. Bonus points if the flow actually works for real (different headers build on what came before, and commands actually work).

Reply:
See above -- we do lead with the "Layer configuration examples" section near the top of the doc. It gives a basic example and describes the replace vs merge override feature.

Re bonus points: I agree it'd be better if this was more real-world and flow-on. However, given that section is already in place, I'm going to leave that for a future PR. I think it's fine to make incremental progress here.

Reply:
Got it, thanks.